### ISSUE TITLE
PythonLexer fails to correctly tokenize a string

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#113]
The `PythonLexer` bails on the following line of valid Python source:
https://github.com/python/cpython/blob/fe0d9e22a52a10c4cbe52254b51f2d4e74d83568/Lib/test/test_unparse.py#L183

which is

```python
        self.check_ast_roundtrip('''f\'\'\'-{f"""*{f"+{f'.{x}.'}+"}*"""}-'single quote\\'\'\'\'''')
```

The failure:
```
$ java PyTest /home/vinay/projects/cpython/Lib/test/test_unparse.py
Processing /home/vinay/projects/cpython/Lib/test/test_unparse.py ... Error processing file: /home/vinay/projects/cpython/Lib/test/test_unparse.py
org.parsers.python.ParseException: 
Encountered an error at (or somewhere around) /home/vinay/projects/cpython/Lib/test/test_unparse.py:183:98
Was expecting one of the following:
RPAREN
Found string "\'" of type INVALID
	at org.parsers.python.PythonParser.handleUnexpectedTokenType(PythonParser.java:28133)
	at org.parsers.python.PythonParser.consumeToken(PythonParser.java:28121)
	at org.parsers.python.PythonParser.InvocationArguments(PythonParser.java:5090)
	at PythonParser.PrimaryExpression(Python.javacc:270)
	at org.parsers.python.PythonParser.PrimaryExpression(PythonParser.java:4632)
	at PythonParser.AwaitPrimary(Python.javacc:259)
	at org.parsers.python.PythonParser.AwaitPrimary(PythonParser.java:4433)
	at PythonParser.Power(Python.javacc:250)
	at org.parsers.python.PythonParser.Power(PythonParser.java:4215)
	at PythonParser.UnaryExpression(Python.javacc:248)
	at org.parsers.python.PythonParser.UnaryExpression(PythonParser.java:4125)
	at PythonParser.MultiplicativeExpression(Python.javacc:246)
	at org.parsers.python.PythonParser.MultiplicativeExpression(PythonParser.java:4024)
	at PythonParser.AdditiveExpression(Python.javacc:244)
	at org.parsers.python.PythonParser.AdditiveExpression(PythonParser.java:3936)
	at PythonParser.ShiftExpression(Python.javacc:242)
	at org.parsers.python.PythonParser.ShiftExpression(PythonParser.java:3848)
	at PythonParser.BitwiseAnd(Python.javacc:240)
	at org.parsers.python.PythonParser.BitwiseAnd(PythonParser.java:3771)
	at PythonParser.BitwiseXor(Python.javacc:238)
	at org.parsers.python.PythonParser.BitwiseXor(PythonParser.java:3694)
	at PythonParser.BitwiseOr(Python.javacc:236)
	at org.parsers.python.PythonParser.BitwiseOr(PythonParser.java:3617)
	at PythonParser.Comparison(Python.javacc:363)
	at org.parsers.python.PythonParser.Comparison(PythonParser.java:6359)
	at PythonParser.Inversion(Python.javacc:361)
	at org.parsers.python.PythonParser.Inversion(PythonParser.java:6304)
	at PythonParser.Conjunction(Python.javacc:359)
	at org.parsers.python.PythonParser.Conjunction(PythonParser.java:6224)
	at PythonParser.Disjunction(Python.javacc:357)
	at org.parsers.python.PythonParser.Disjunction(PythonParser.java:6147)
	at PythonParser.Expression(Python.javacc:355)
	at org.parsers.python.PythonParser.Expression(PythonParser.java:6047)
	at PythonParser.StarExpression(Python.javacc:229)
	at org.parsers.python.PythonParser.StarExpression(PythonParser.java:3389)
	at PythonParser.StarExpressions(Python.javacc:232)
	at org.parsers.python.PythonParser.StarExpressions(PythonParser.java:3449)
	at PythonParser.SmallStatement(Python.javacc:87)
	at org.parsers.python.PythonParser.SmallStatement(PythonParser.java:1168)
	at PythonParser.SimpleStatement(Python.javacc:82)
	at org.parsers.python.PythonParser.SimpleStatement(PythonParser.java:1076)
	at PythonParser.Statement(Python.javacc:77)
	at org.parsers.python.PythonParser.Statement(PythonParser.java:1009)
	at PythonParser.Block(Python.javacc:153)
	at org.parsers.python.PythonParser.Block(PythonParser.java:1845)
	at PythonParser.FunctionDefinition(Python.javacc:346)
	at org.parsers.python.PythonParser.FunctionDefinition(PythonParser.java:5927)
	at PythonParser.CompoundStatement(Python.javacc:113)
	at org.parsers.python.PythonParser.CompoundStatement(PythonParser.java:1329)
	at PythonParser.Statement(Python.javacc:79)
	at org.parsers.python.PythonParser.Statement(PythonParser.java:1020)
	at PythonParser.Block(Python.javacc:153)
	at org.parsers.python.PythonParser.Block(PythonParser.java:1845)
	at PythonParser.ClassDefinition(Python.javacc:334)
	at org.parsers.python.PythonParser.ClassDefinition(PythonParser.java:5818)
	at PythonParser.CompoundStatement(Python.javacc:117)
	at org.parsers.python.PythonParser.CompoundStatement(PythonParser.java:1351)
	at PythonParser.Statement(Python.javacc:79)
	at org.parsers.python.PythonParser.Statement(PythonParser.java:1020)
	at PythonParser.Module(Python.javacc:25)
	at org.parsers.python.PythonParser.Module(PythonParser.java:351)
failed.
0 successes, 1 failures, 1.47 secs.
Failures:
  /home/vinay/projects/cpython/Lib/test/test_unparse.py
```
It's very similar to a string on the preceding line, which is parsed correctly. It's not actually an f-string, though the failure occurs in a method called `test_fstrings_complicated`.

From what I can see, it's a normal string starting and ending with `'''` but which contains 3 escaped single quotes just before the end.

**Comment by vsajip** [comment]
Could this be the problem?
https://github.com/javacc21/javacc21/blob/1cba7c4978838c2255945e2fdb828e14a3f7b1a3/examples/python/PythonLexer.javacc#L349
I'm not sure what's meant by `~[]` in that rule. Seems like it should be more explicit, looking for a/b/n/f/r/t/v or a hex or Unicode escape sequence.

**Comment by revusky** [comment]
> I'm not sure what's meant by `~[]` in that rule. Seems like it should be more explicit, looking for a/b/n/f/r/t/v or a hex or Unicode escape sequence.

No, actually, that line is not a problem. The `~[]` means that we match any character at all, which is what you get if you take the complement of the empty character set. And that, in fact, is what the string escape is here, the backslash followed by any character at all. 

The problem was elsewhere. I had written the regexp for those multiline string literals incorrectly. The way I had written it, if you had something like: `''''\''''` that would not work because it would scan the `'` character followed by the `\` and then scan 3 single quotes to terminate and there would still be one single-quote at the end dangling. So in these cases, it needs to scan the lone single-quote followed by the STRING_ESCAPE and then it gets to the triple-quote at the end. I rewrote it more clearly by separating out a little helper regexp called INSIDE_SL_3. In general, it seems that when a regexp gets too hairy, one really has to break it out sub-expressions. Of course, that's true of coding generally, I suppose. If a method gets too big, then....

Well, I don't know if that explanation is clear, but if you look at how it's expressed now and compare it to what was there before...

Oh, another thing is that this bug had nothing to do with f-strings in particular. That these are f-strings is incidental. If the 'f' was not there, the bug would still be there. As regards f-strings, I was looking at that a bit more, in particular https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals

I had never looked at that closely before, but if we wanted to validate the f-strings, I guess that needs to happen in a second pass, like walk the tree, find all the f-strings and validate them separately. I mean, this makes reference to things in the syntactic grammar. If we're handling the string literals, including f-strings just lexically, then that's at a lower level of analysis, and we can't refer to the syntactic grammar when we're just lexing, you understand. But I don't think that validating the f-strings is a very urgent matter at the moment.

**Comment by vsajip** [comment]
Should be fixed in 151966f.
